lightning/backend/local.go

```diff
@@ -177,17 +186,122 @@ func (e *LocalFile) getSizeProperties() (*sizeProperties, error) {
 	return sizeProps, nil
 }
 
+func (e *LocalFile) isLocked() bool {
+	return e.isImportingAtomic.Load() != 0
+}
+
+func (e *LocalFile) getEngineFileSize() EngineFileSize {
+	metrics := e.db.Metrics()
+	total := metrics.Total()
+	var memSize int64
+	e.localWriters.Range(func(k, v interface{}) bool {
+		w := k.(*LocalWriter)
+		memSize += w.writeBatch.totalSize
+		if w.writer != nil {
+			total.Size += int64(w.writer.writer.EstimatedSize())
+		}
+		return true
+	})
+
+	return EngineFileSize{
+		UUID:        e.Uuid,
+		DiskSize:    total.Size,
+		MemSize:     memSize,
+		IsImporting: e.isLocked(),
+	}
+}
+
 // lock locks the local file for importing.
 func (e *LocalFile) lock(state importMutexState) {
 	e.mutex.Lock()
-	atomic.StoreInt32(&e.isImportingAtomic, int32(state))
+	e.isImportingAtomic.Store(uint32(state))
+}
+
+// lockUnless tries to lock the local file unless it is already locked into the state given by
+// ignoreStateMask. Returns whether the lock is successful.
+func (e *LocalFile) lockUnless(newState, ignoreStateMask importMutexState) bool {
+	curState := e.isImportingAtomic.Load()
+	if curState&uint32(ignoreStateMask) != 0 {
+		return false
+	}
+	e.lock(newState)
+	return true
 }
 
 func (e *LocalFile) unlock() {
-	atomic.StoreInt32(&e.isImportingAtomic, int32(importMutexStateNoLock))
+	if e == nil {
+		return
+	}
+	e.isImportingAtomic.Store(0)
 	e.mutex.Unlock()
 }
 
+func (e *LocalFile) flushLocalWriters(parentCtx context.Context) error {
+	eg, ctx := errgroup.WithContext(parentCtx)
+	e.localWriters.Range(func(k, v interface{}) bool {
+		eg.Go(func() error {
+			w := k.(*LocalWriter)
+			replyErrCh := make(chan error, 1)
+			w.flushChMutex.RLock()
+			if w.flushCh != nil {
+				w.flushCh <- replyErrCh
+			} else {
+				replyErrCh <- nil
+			}
+			w.flushChMutex.RUnlock()
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case err := <-replyErrCh:
+				return err
+			}
+		})
+		return true
+	})
+	return eg.Wait()
+}
+
+func (e *LocalFile) flushEngineUnlocked(ctx context.Context) error {
```

Comment:
not understand what "Unlocked" means. This method is called with engine file locked

Reply:
it means this function itself has no locking. any better name 🤗

Reply:
no good idea 😂 maybe we could remove "Unlocked". I can't think out a name clearly indicating locking is prerequisite or locking is behaviour of the function. WDYT @glorv @overvenus @Little-Wallace